File: backend/database/deploy_procedures.js
/**
 * HRMS Stored Procedures Deployment Script (Node.js)
 * ====================================================
 * HOW TO RUN:
 *   npm install mysql2          (from project root)
 *   cd backend\database
 *   node deploy_procedures.js
 */

const mysql = require('mysql2/promise');
const fs    = require('fs');
const path  = require('path');

const DB_CONFIG = {
  host:     'hrms.c02gczm2cgx8.us-east-1.rds.amazonaws.com',
  port:     4306,
  user:     'HRMSadmin',
  password: 'HwULGJ6gbxQIhzwNeZ9L',
  database: 'hrms_db',
  multipleStatements: false,
  connectTimeout: 30000,
};

const SQL_FILE = path.join(__dirname, 'all_procedures.sql');

function parseStatements(sqlContent) {
  // The file uses $$ as terminator (no DELIMITER declaration at top).
  // Split the whole file on $$ to get individual statements.
  const chunks = sqlContent.split('$$');
  const statements = [];

  for (const chunk of chunks) {
    const stmt = chunk.trim();
    if (!stmt || stmt.startsWith('--') || stmt.length < 10) continue;

    // Remove standalone USE statements (we're already on the right DB)
    if (/^\s*USE\s+\w+\s*;?\s*$/i.test(stmt)) continue;

    // Remove pure comment blocks
    const noComments = stmt.replace(/--[^\n]*/g, '').trim();
    if (!noComments) continue;

    statements.push(stmt);
  }

  return statements;
}

async function main() {
  if (!fs.existsSync(SQL_FILE)) {
    console.error(`ERROR: ${SQL_FILE} not found.`);
    process.exit(1);
  }

  const sqlContent = fs.readFileSync(SQL_FILE, 'utf8');
  console.log(`Read all_procedures.sql (${sqlContent.length.toLocaleString()} bytes)`);

  const statements = parseStatements(sqlContent);
  console.log(`Found ${statements.length} statements to execute\n`);

  console.log(`Connecting to ${DB_CONFIG.host}:${DB_CONFIG.port} ...`);
  let conn;
  try {
    conn = await mysql.createConnection(DB_CONFIG);
    console.log('Connected!\n');
  } catch (err) {
    console.error(`Connection failed: ${err.message}`);
    process.exit(1);
  }

  let success = 0, errors = 0;
  const failed = [];

  for (let i = 0; i < statements.length; i++) {
    const stmt = statements[i];
    try {
      await conn.query(stmt);
      const m = stmt.match(/CREATE\s+PROCEDURE\s+(\w+)/i);
      if (m) {
        console.log(`  [${String(i+1).padStart(3)}] OK  ${m[1]}`);
        success++;
      }
    } catch (err) {
      const m = stmt.match(/CREATE\s+PROCEDURE\s+(\w+)/i) ||
                stmt.match(/DROP\s+PROCEDURE.*?(\w+)\s*$/i);
      const name = m ? m[1] : `stmt_${i+1}`;
      // Only log CREATE errors (DROP IF EXISTS errors are harmless)
      if (/CREATE\s+PROCEDURE/i.test(stmt)) {
        console.log(`  [${String(i+1).padStart(3)}] ERR ${name}: ${err.message}`);
        failed.push({ name, error: err.message });
        errors++;
      }
    }
  }

  // Verify count
  let dbCount = '?';
  try {
    const [rows] = await conn.execute(
      "SELECT COUNT(*) AS cnt FROM information_schema.ROUTINES WHERE ROUTINE_SCHEMA = ? AND ROUTINE_TYPE = 'PROCEDURE'",
      [DB_CONFIG.database]
    );
    dbCount = rows[0].cnt;
  } catch (_) {}

  await conn.end();

  console.log('\n' + '='.repeat(60));
  console.log('DEPLOYMENT COMPLETE');
  console.log(`  Procedures created successfully : ${success}`);
  console.log(`  Errors                         : ${errors}`);
  console.log(`  Procedures in DB (verified)    : ${dbCount}`);
  console.log('='.repeat(60));

  if (failed.length) {
    console.log(`\nFailed (${failed.length}):`);
    failed.forEach(f => console.log(`  - ${f.name}: ${f.error}`));
  }
}

main().catch(err => { console.error(err); process.exit(1); });

File: backend/database/deploy_procedures.py
"""
HRMS Stored Procedures Deployment Script
=========================================
Deploys all 234 stored procedures to AWS RDS MySQL.

HOW TO RUN (from Windows Command Prompt or PowerShell):
  1. Install PyMySQL if not installed:
       pip install pymysql

  2. Run this script from the backend/database folder:
       cd C:\TimeSheet\humanresourceshradmintemplate\backend\database
       python deploy_procedures.py
"""

import re
import sys

# ── Connection Settings ────────────────────────────────────────────────────────
HOST     = "hrms.c02gczm2cgx8.us-east-1.rds.amazonaws.com"
PORT     = 4306
USER     = "HRMSadmin"
PASSWORD = "HwULGJ6gbxQIhzwNeZ9L"
DATABASE = "hrms_db"
SQL_FILE = "all_procedures.sql"
# ──────────────────────────────────────────────────────────────────────────────

def parse_procedures(sql_content):
    """Parse SQL file with DELIMITER $$ syntax into individual statements."""
    statements = []
    current_delimiter = ";"
    current_stmt = []

    for line in sql_content.splitlines():
        stripped = line.strip()
        delimiter_match = re.match(r'DELIMITER\s+(\S+)', stripped, re.IGNORECASE)
        if delimiter_match:
            current_delimiter = delimiter_match.group(1)
            continue

        if stripped.endswith(current_delimiter):
            line_without_delim = line.rstrip()
            if line_without_delim.endswith(current_delimiter):
                line_without_delim = line_without_delim[:-len(current_delimiter)]
            current_stmt.append(line_without_delim)
            stmt = "\n".join(current_stmt).strip()
            if stmt:
                statements.append(stmt)
            current_stmt = []
        else:
            current_stmt.append(line)

    remaining = "\n".join(current_stmt).strip()
    if remaining:
        statements.append(remaining)

    return statements


def main():
    try:
        import pymysql
    except ImportError:
        print("ERROR: PyMySQL is not installed.")
        print("Run:  pip install pymysql")
        sys.exit(1)

    try:
        with open(SQL_FILE, "r", encoding="utf-8") as f:
            sql_content = f.read()
        print(f"Read {SQL_FILE} ({len(sql_content):,} bytes)")
    except FileNotFoundError:
        print(f"ERROR: {SQL_FILE} not found.")
        print("Make sure you are running from the backend/database directory.")
        sys.exit(1)

    statements = parse_procedures(sql_content)
    statements = [s for s in statements if len(s.strip()) > 10]
    print(f"Parsed {len(statements)} statements")

    print(f"\nConnecting to {HOST}:{PORT} ...")
    try:
        conn = pymysql.connect(
            host=HOST, port=PORT, user=USER, password=PASSWORD,
            database=DATABASE, charset="utf8mb4", connect_timeout=30,
        )
        print("Connected successfully!\n")
    except Exception as e:
        print(f"ERROR: Could not connect: {e}")
        sys.exit(1)

    cursor = conn.cursor()
    success_count = 0
    error_count = 0
    errors = []

    for i, stmt in enumerate(statements, 1):
        if re.match(r'^\s*USE\s+', stmt, re.IGNORECASE):
            continue
        cleaned = re.sub(r'--[^\n]*', '', stmt).strip()
        if not cleaned:
            continue

        try:
            cursor.execute(stmt)
            conn.commit()
            name_match = re.search(r'CREATE\s+PROCEDURE\s+(\w+)', stmt, re.IGNORECASE)
            if name_match:
                print(f"  [{i:3d}] OK  {name_match.group(1)}")
                success_count += 1
        except Exception as e:
            name_match = re.search(r'CREATE\s+PROCEDURE\s+(\w+)', stmt, re.IGNORECASE)
            proc_name = name_match.group(1) if name_match else f"stmt_{i}"
            print(f"  [{i:3d}] ERR {proc_name}: {e}")
            errors.append((proc_name, str(e)))
            error_count += 1

    cursor.close()

    try:
        verify_conn = pymysql.connect(
            host=HOST, port=PORT, user=USER, password=PASSWORD,
            database=DATABASE, charset="utf8mb4"
        )
        vcur = verify_conn.cursor()
        vcur.execute(
            "SELECT COUNT(*) FROM information_schema.ROUTINES "
            "WHERE ROUTINE_SCHEMA = %s AND ROUTINE_TYPE = 'PROCEDURE'",
            (DATABASE,)
        )
        db_count = vcur.fetchone()[0]
        vcur.close()
        verify_conn.close()
    except Exception:
        db_count = "unknown"

    print("\n" + "=" * 60)
    print(f"DEPLOYMENT COMPLETE")
    print(f"  Procedures deployed successfully : {success_count}")
    print(f"  Errors                          : {error_count}")
    print(f"  Procedures in DB (verified)     : {db_count}")
    print("=" * 60)

    if errors:
        print(f"\nFailed procedures ({len(errors)}):")
        for name, err in errors:
            print(f"  - {name}: {err}")

    conn.close()


if __name__ == "__main__":
    main()
